pageserver/src/tenant.rs

```diff
@@ -997,16 +1145,46 @@ impl Tenant {
             }
         }
 
+        let timelines_to_load = match load_reason {
+            TenantLoadReason::Regular => timelines_to_load
+                .into_iter()
+                .map(|(timeline_id, metadata)| {
+                    (timeline_id, (metadata, TimelineLoadReason::Regular))
+                })
+                .collect(),
+            TenantLoadReason::Attaching {
+                mut attach_data_by_timeline_id,
+            } => {
+                let mut merged = HashMap::new();
```

Comment:
```suggestion
                let mut merged = HashMap::with_capacity(timelines_to_load.len());
```

Alternative is to turn the `for (_, _) in timelines.to_load.drain()` to a iterator like which also carries the size hint used for `with_capacity`:

```rust
timelines_to_load.drain()
  .map(|(id, meta)| {
    let val = attach_data_by_timeline_id.remove(&id).context("...")?;
    let val = TimelineLoadReason::Attach(data);
    Ok((id, (meta, val)))
  })
  .collect::<Result<HashMap<_, _>, _>()?;
```

(The `FromIterator` impl used here is better explained in doc comment: https://doc.rust-lang.org/stable/src/core/result.rs.html#2030-2077)

This would actually make it more clear that:
- `assert!(replaced.is_none())` -- the two maps `merged` and `timelines_to_laod` both have the same key, impossible for this to happen